File: DynamicKey/AgoraDynamicKey/python/src/SimpleTokenBuilder.py
import os
import sys
from collections import OrderedDict

sys.path.append(os.path.join(os.path.dirname(__file__), '../src'))
from AccessToken import *

Role_Attendee = 1
Role_Publisher = 2
Role_Subscriber = 3
Role_Admin = 4

AttendeePrivileges = OrderedDict([
    (kJoinChannel, 0),
    (kPublishAudioStream, 0),
    (kPublishVideoStream, 0),
    (kPublishDataStream, 0)
])

PublisherPrivileges = OrderedDict([
    (kJoinChannel, 0),
    (kPublishAudioStream, 0),
    (kPublishVideoStream, 0),
    (kPublishDataStream, 0),
    (kPublishAudiocdn, 0),
    (kPublishVideoCdn, 0),
    (kInvitePublishAudioStream, 0),
    (kInvitePublishVideoStream, 0),
    (kInvitePublishDataStream, 0)
])

SubscriberPrivileges = OrderedDict([
    (kJoinChannel, 0),
    (kRequestPublishAudioStream, 0),
    (kRequestPublishVideoStream, 0),
    (kRequestPublishDataStream, 0)
])

AdminPrivileges = OrderedDict([
    (kJoinChannel, 0),
    (kPublishAudioStream, 0),
    (kPublishVideoStream, 0),
    (kPublishDataStream, 0),
    (kAdministrateChannel, 0)
])

RolePrivileges = OrderedDict([
    (Role_Attendee, AttendeePrivileges),
    (Role_Publisher, PublisherPrivileges),
    (Role_Subscriber, SubscriberPrivileges),
    (Role_Admin, AdminPrivileges)
])


class SimpleTokenBuilder:

    def __init__(self, appID, appCertificate, channelName, uid):
        self.token = AccessToken(appID, appCertificate, channelName, uid)

    def initPrivileges(self, role):
        self.token.messages = RolePrivileges[role]

    def initTokenBuilder(self, originToken):
        return self.token.fromString(originToken)

    def setPrivilege(self, privilege, expireTimestamp):
        self.token.messages[privilege] = expireTimestamp

    def removePrivilege(self, privilege):
        self.token.messages.pop(privilege)

    def buildToken(self):
        return self.token.build()

File: DynamicKey/AgoraDynamicKey/nodejs/src/SimpleTokenBuilder.js
var AccessToken = require('../src/AccessToken').AccessToken;
var Priviledges = require('../src/AccessToken').priviledges;

var Role = {
    kRoleAttendee: 0,  // for communication
    kRolePublisher: 1, // for live broadcast
    kRoleSubscriber: 2,  // for live broadcast
    kRoleAdmin: 101
};

var attendeePrivileges = {
    [Priviledges.kJoinChannel]: 0,
    [Priviledges.kPublishAudioStream]: 0,
    [Priviledges.kPublishVideoStream]: 0,
    [Priviledges.kPublishDataStream]: 0
};


var publisherPrivileges = {
    [Priviledges.kJoinChannel]: 0,
    [Priviledges.kPublishAudioStream]: 0,
    [Priviledges.kPublishVideoStream]: 0,
    [Priviledges.kPublishDataStream]: 0,
    [Priviledges.kPublishAudiocdn]: 0,
    [Priviledges.kPublishVideoCdn]: 0,
    [Priviledges.kInvitePublishAudioStream]: 0,
    [Priviledges.kInvitePublishVideoStream]: 0,
    [Priviledges.kInvitePublishDataStream]: 0
};

var subscriberPrivileges = {
    [Priviledges.kJoinChannel] : 0,
    [Priviledges.kRequestPublishAudioStream] : 0,
    [Priviledges.kRequestPublishVideoStream] : 0,
    [Priviledges.kRequestPublishDataStream] : 0
};

var adminPrivileges = {
    [Priviledges.kJoinChannel] : 0,
    [Priviledges.kPublishAudioStream] : 0,
    [Priviledges.kPublishVideoStream] : 0,
    [Priviledges.kPublishDataStream] : 0,
    [Priviledges.kAdministrateChannel]: 0
};

var RolePrivileges = {
	[Role.kRoleAttendee]: attendeePrivileges,
    [Role.kRolePublisher]: publisherPrivileges,
    [Role.kRoleSubscriber]: subscriberPrivileges,
    [Role.kRoleAdmin]: adminPrivileges
}

var SimpleTokenBuilder = function (appID, appCertificate, channelName, uid) {
    let builder = this;
    this.key = new AccessToken(appID, appCertificate, channelName, uid);

    this.buildToken = function () {
        return builder.key.build();
    }

    this.initPrivileges = function (role) {
        let rolePri = RolePrivileges[role];
        builder.key.messages = JSON.parse(JSON.stringify(rolePri));
    }

    this.initTokenBuilder = function (originToken) {
        return builder.key.fromString(originToken);
    }

    this.setPrivilege = function (privilege, expireTimestamp) {
        builder.key.messages[privilege] = expireTimestamp;
    }
    
    this.removePrivilege = function (privilege) {
        delete builder.key.messages[privilege];
    }
};

module.exports.SimpleTokenBuilder = SimpleTokenBuilder;
module.exports.Role = Role;